src/dbnode/storage/index.go

```diff
@@ -716,77 +723,196 @@ func (i *nsIndex) Query(
+	ctx context.Context,
+) time.Duration {
+	// TODO(r): Allow individual queries to specify timeouts using
+	// deadlines passed by the context
+	return i.state.runtimeOpts.defaultQueryTimeout
+}
+
+func (i *nsIndex) overriddenOptsForQueryWithLock(
+	opts index.QueryOptions,
+) index.QueryOptions {
+	// Override query response limit if needed
 	if i.state.runtimeOpts.maxQueryLimit > 0 && (opts.Limit == 0 ||
 		int64(opts.Limit) > i.state.runtimeOpts.maxQueryLimit) {
 		i.logger.Debugf("overriding query response limit, requested: %d, max-allowed: %d",
 			opts.Limit, i.state.runtimeOpts.maxQueryLimit) // FOLLOWUP(prateek): log query too once it's serializable.
 		opts.Limit = int(i.state.runtimeOpts.maxQueryLimit)
 	}
+	return opts
+}
 
-	var (
-		exhaustive = true
-		results    = i.opts.IndexOptions().ResultsPool().Get()
-		err        error
-	)
-	results.Reset(i.nsMetadata.ID())
-	ctx.RegisterFinalizer(results)
-
+func (i *nsIndex) blocksForQueryWithLock(queryRange xtime.Ranges) ([]index.Block, error) {
 	// Chunk the query request into bounds based on applicable blocks and
 	// execute the requests to each of them; and merge results.
-	queryRange := xtime.NewRanges(xtime.Range{
-		Start: opts.StartInclusive, End: opts.EndExclusive})
+	blocks := make([]index.Block, 0, len(i.state.blockStartsDescOrder))
 
 	// iterate known blocks in a defined order of time (newest first) to enforce
 	// some determinism about the results returned.
 	for _, start := range i.state.blockStartsDescOrder {
 		block, ok := i.state.blocksByTime[start]
 		if !ok { // should never happen
-			return index.QueryResults{}, i.missingBlockInvariantError(start)
+			return nil, i.missingBlockInvariantError(start)
 		}
 
-		// ensure the block has data requested by the query
+		if queryRange.IsEmpty() {
+			// Terminate if queryRange doesn't need any more data
+			break
+		}
+
+		// Ensure the block has data requested by the query
 		blockRange := xtime.Range{Start: block.StartTime(), End: block.EndTime()}
 		if !queryRange.Overlaps(blockRange) {
 			continue
 		}
 
-		// terminate early if we know we don't need any more results
-		if opts.Limit > 0 && results.Size() >= opts.Limit {
-			exhaustive = false
-			break
-		}
+		// Remove this range from the query range
+		queryRange = queryRange.RemoveRange(blockRange)
 
-		exhaustive, err = block.Query(query, opts, results)
-		if err != nil {
-			return index.QueryResults{}, err
-		}
+		blocks = append(blocks, block)
+	}
 
-		if !exhaustive {
-			// i.e. block had more data but we stopped early, we know
-			// we have hit the limit and don't need to query any more.
-			break
-		}
+	return blocks, nil
+}
 
-		// terminate if queryRange doesn't need any more data
-		queryRange = queryRange.RemoveRange(blockRange)
-		if queryRange.IsEmpty() {
-			break
-		}
-	}
+func (i *nsIndex) newConcurrentResults(ctx context.Context) *index.ConcurrentResults {
+	results := i.opts.IndexOptions().ResultsPool().Get()
```

Comment:
want to just throw this pool onto the struct itself like we do with the worker pool so you don't have to do the triple function call?